### PR TITLE
Thread view

### PR DESCRIPTION
**Pull request by andrew**
⚠️ Work in progress ⚠️ 

Adding a view to see all the comments on an issue/pull request if subject syncing is enabled, as discussed with @csexton in [gitter](https://gitter.im/octobox/octobox)

screen-shot-2018-07-25-at-11 43 23

Still quite a lot to do here, a few things of note:

- [ ] collapse comments that have already been seen (anything before `last_read_at`)
- [ ] Allow replies directly from that view
- [ ] Hook up the archive, star and mute buttons to actually work
- [x] ensure the visuals work correctly with the bootstrap 4 that's currently in progress https://github.com/octobox/octobox/pull/366
- [ ] Tests!
- [ ] Make subject title clickable

This also allows us to show all the people involved in an issue/pr and show them on the main page (cc @chrisarcand as he mentioned something about that data on twitter the other day):

screen_shot_2018-07-25_at_16 24 30

Syncing speed can also be greatly improved by using a similar approach to https://github.com/octobox/octobox/pull/482

### CODE REVIEW

**Comment by chrisarcand** [comment]
Nice, really like where this is going.

**Comment by andrew** [comment]
This should enable https://github.com/octobox/octobox/issues/453 as well

**Comment by andrew** [comment]
This PR needs to be updated to bootstrap-4 now that #366 is merged

**Comment by betatim** [comment]
Will this change how the badges in the second to last column work as well? In particular only showing "Mention" if the mention happened in a comment since `last_read_at`. If it is out of scope for this PR I can create an issue to track this (unless it is out of scope for the project/technically super hard to do).

**Comment by andrew** [comment]
@betatim this pr will add the data to make that possible, definitely open a new issue for that particular feature request so it doesn't get lost in this already big feature.

**Comment by andrew** [comment]
Current visuals:

pasted_image_at_2018-08-03__4_40_pm

**Comment by csexton** [comment]
Playing around with this on my local copy of the branch it didn't look the same as your screenshot. I know this is a WIP, but wanted to follow along on the development. Two things I see:

1) When click on an issue I am redirected to a JSON page on GitHub's API. In other words the `notification-subject` link is to a page like [this](https://api.github.com/repos/octobox/octobox/pulls/709).

2) Show notification page looks different than your screen shot. If I edit the URL directly I can see the page, but it looks something like this:

image

Per your request on the gitters:

<details>
  <summary>My .env file</summary>

```sh
DATABASE=postgres
GITHUB_CLIENT_ID=REDACTED
GITHUB_CLIENT_SECRET=REDACTED
```

</details>

**Comment by andrew** [comment]
@csexton you'll need to enable subject syncing to get it to download all the data for the pull request and comments, add `FETCH_SUBJECT=true` to your `.env` file, restart the app, log back out and in again and then resync.

**Comment by csexton** [comment]
Thanks, it's beautiful!

image

**Comment by pradyunsg** [comment]
It'd be nice if reactions could be automatically collapsed and if clicking on the issue/PR title takes you to the GitHub thread.

**Comment by andrew** [comment]
@pradyunsg reactions probably won't make it into this PR so we can visit that later and +1 on making the subject title clickable

**Comment by andrew** [comment]
@BenJam what’s left to do on this?

**Comment by BenJam** [comment]
not much...

- mark comments as read when you scan down the page (in JS)
- collapse comments when read, with down arrow to uncollapse
- tests 😬

**Comment by pradyunsg** [comment]
Just a thought: Having a split pane view so that when you click on a notification it opens on the side. Would be super neat for big screens (and ultrawide ones). I guess this should be user configurable though and it's substantially more complex than simply having it be a separate view.

screen shot 2018-08-16 at 3 18 32 pm


PS: Let me know if me suggesting features/ideas like this gets annoying or something.

**Comment by andrew** [comment]
@pradyunsg your suggestions are definitely not annoying, keep them coming!

For our first pass at this I think it's a small bit of work to keep it as a separate page, we can then revisit adding a split pane view for larger screens in a later pr.

File: app/helpers/notifications_helper.rb [hunk 1]
```diff
@@ -82,6 +82,10 @@ def filtered_params(override = {})
     filters.merge(override)
   end
 
+  def mute_button()
```
Comment (andrew) [inline app/helpers/notifications_helper.rb]: No need for empty brackets on these methods if they don't take arguments

**Comment by BenJam** [comment]
latest:
screen shot 2018-08-17 at 19 56 28

The animations overlapping are killing me atm. Need to fiddle with a JS animation to get the ordering right as BS just isn't working well enough. Anyone who can get it working without that feel free to drop it 👍

**Comment by pradyunsg** [comment]
Nit: This does not render in a GFM compatible manner.

**Comment by andrew** [comment]
@pradyunsg can you be more specific about which bits of gfm are missing?

**Comment by pradyunsg** [comment]
In the screenshot @BenJam posted above, the checklist isn't rendered as a checklist.

---

I just deployed this to my personal instance to try it out. Some thoughts from what I noticed in the 4 notifications I have currently:

- Would it be fine to not set `font-size: 16px` in markdown-body -- it makes the text of the thread larger than the text of the sidebar; which I found a little odd on first glance.
- This PR seems to be missing description from first comments on issues:

screen shot 2018-08-18 at 11 57 35 am
screen shot 2018-08-18 at 11 57 46 am

- The full width of the comment bodies, make it substantially more difficult to read. Along the lines of #793, except much more pronounced due to a larger amount of text that is meant to be read.
- Non-comment events are missing; which makes the view not-useful for PRs.
-

**Comment by pradyunsg** [comment]
The missing descriptions seem odd; since they're clearly a part of @andrew's and @BenJam's screenshots above. :/

**Comment by pradyunsg** [comment]
More thoughts, related to markdown rendering:

screen shot 2018-08-18 at 8 01 09 pm

- Linkify http[s] URLs
- Linkify mentions, to teams/organizations/users. (I'm not sure how feasible this is)


Further, I just had this PR fail on me:

<details>

```
2018-08-18T14:34:28.886061+00:00 app[web.1]: I, [2018-08-18T14:34:28.885932 #20]  INFO -- : [a7bf9503-f216-45de-8deb-74a33204da12] method=POST path=/notifications/505/mark_read format=*/* controller=NotificationsController action=mark_read status=200 duration=10.17 view=0.00 db=4.73
2018-08-18T14:34:29.073406+00:00 app[web.1]: I, [2018-08-18T14:34:29.073285 #15]  INFO -- : [cba758f0-311b-4b71-9914-ab93bb5ccb59] method=GET path=/notifications/505 format=html controller=NotificationsController action=show status=404 error='ActiveRecord::RecordNotFound: Couldn't find Notification with 'id'=505 [WHERE "notifications"."user_id" = $1 AND "notifications"."archived" = $2 AND "notifications"."unread" = $3]' duration=15.23 view=0.00 db=8.27
2018-08-18T14:34:29.074001+00:00 app[web.1]: F, [2018-08-18T14:34:29.073939 #15] FATAL -- : [cba758f0-311b-4b71-9914-ab93bb5ccb59]
2018-08-18T14:34:29.074064+00:00 app[web.1]: F, [2018-08-18T14:34:29.074010 #15] FATAL -- : [cba758f0-311b-4b71-9914-ab93bb5ccb59] ActiveRecord::RecordNotFound (Couldn't find Notification with 'id'=505 [WHERE "notifications"."user_id" = $1 AND "notifications"."archived" = $2 AND "notifications"."unread" = $3]):
2018-08-18T14:34:29.074123+00:00 app[web.1]: F, [2018-08-18T14:34:29.074066 #15] FATAL -- : [cba758f0-311b-4b71-9914-ab93bb5ccb59]
2018-08-18T14:34:29.074167+00:00 app[web.1]: F, [2018-08-18T14:34:29.074120 #15] FATAL -- : [cba758f0-311b-4b71-9914-ab93bb5ccb59] app/controllers/notifications_controller.rb:97:in `show'
```

</details>

**Comment by pradyunsg** [comment]
screen shot 2018-08-18 at 8 12 13 pm

The "pull request opened" datetime is being mixed with notification generated datatime, in the above screenshot. Also, minor nits: space in "pullrequest". And and, the colours are misleading -- to me, it suggested "opened this pullrequest" is clickable and "octobox/octobox" isn't, while the exact opposite is true.

(I understand that this is a WIP PR; so please let me know if any of what I say is annoying or incorrect etc 🙃)

**Comment by andrew** [comment]
It's all helpful feedback @pradyunsg, thanks for the extra set of 👀 on this

**Comment by andrew** [comment]
Once #806 is merged we'll be able to listen for [`issue_comment`](https://developer.github.com/v3/activity/events/types/#issuecommentevent) web hooks and update the threads instantly.

**Comment by andrew** [comment]
Another potential feature we can enable with this PR is the "reopened" status and icon, as discussed in https://github.com/octobox/octobox/pull/827#issuecomment-414740677

**Comment by BenJam** [comment]
I'm jumping back into this now that #1040 and #1030 have made things a little easier. Big, scary merge first.

**Comment by andrew** [comment]
This branch is being pulled apart to make it a little easier to deal with, it's been broken up into three parts:

- https://github.com/octobox/octobox/pull/1095
- https://github.com/octobox/octobox/pull/1094
- https://github.com/octobox/octobox/tree/thread-view-2

**Pull request closed without merging**